### PR TITLE
#1101 Allow users to reset their password

### LINKED ISSUE #1101: Let users reset/change their password
As far as I know, there is no mechanism right now for changing your password. And if a user forgets their password, there is no way for them to recover their account. Seems like a pretty basic thing that any account-holding user would expect to have.

### PR DESCRIPTION
**Pull request by ThatOneGoat**
Allow users to reset their password for their Project Sidewalk account. 

The reset pipeline works as follows: From either the sign-in page or sign-in modal, the user can access the "forgot password" page from a link at the bottom of the sign-in form. They will be redirected to the "forgot password" page and then prompted to enter their email. From the email the user entered, a temporary authentication token will be created which will allow the user to access the "reset password" page. An email is sent to the user and they can access the "reset password" page from the link in the email. Temporary authentication tokens are cleaned up every 5 minutes from the database. 

Other changes that are relevant to how the "reset password" pipeline flows: We used to store two user-login_infos for users who signed up (one for their anon account before signing up and the other their actual account). Now, whenever a user signs up, it justs updates the user_login_info for their user_id. Duplicate entries for user_login_info have been cleared.

Resolves #1101

### CODE REVIEW

**Comment by ThatOneGoat** [comment]
I think small safety features like making sure users can only reset their password once every "x" day can be implemented later in a different PR (this one's kinda bulky).

We should also decide on whether the "Project Sidewalk" email should be no-reply or interactable. Most sources say that "no-reply" is kinda alienating to users so maybe we could have the new Project Sidewalk email as some sort of help desk kind of thing?

File: app/models/services/AuthTokenServiceImpl.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import play.api.libs.concurrent.Execution.Implicits._
+
+/**
+ * Handles actions to auth tokens.
+ *
+ * @param authTokenDAO The auth token DAO implementation.
+ * @param clock        The clock instance.
```
Comment (misaugstad) [inline app/models/services/AuthTokenServiceImpl.scala:22]: Are either of these parameters used in this file?

File: app/utils/di/SilhouetteModule.scala [hunk 2]
```diff
@@ -35,13 +35,17 @@ class SilhouetteModule extends AbstractModule with ScalaModule {
...
+    //bind[AuthTokenDAO].to[AuthTokenDAOImpl]
+    bind[AuthTokenDAO].to[AuthTokenDAOSlick]
     bind[DelegableAuthInfoDAO[PasswordInfo]].to[PasswordInfoDAOSlick]
     bind[CacheLayer].to[PlayCacheLayer]
     bind[HTTPLayer].to[PlayHTTPLayer]
     bind[IDGenerator].toInstance(new SecureRandomIDGenerator())
     bind[PasswordHasher].toInstance(new BCryptPasswordHasher)
     bind[FingerprintGenerator].toInstance(new DefaultFingerprintGenerator(false))
     bind[EventBus].toInstance(EventBus())
+    bind[Clock].toInstance(Clock())
```
Comment (misaugstad) [inline app/utils/di/SilhouetteModule.scala:48]: Are we using this anywhere?

File: app/views/forgotPassword.scala.html [hunk 3]
```diff
@@ -0,0 +1,18 @@
+@(forgotPasswordForm: Form[String])(implicit request: RequestHeader, lang: Lang)
+@import views.html.bootstrap._
+
+@main("Project Sidewalk - Recover Password") {
+    @navbar(None)
+    <fieldset class="col-md-6 col-md-offset-3" style="margin-top:51px">
+        <legend>@Messages("forgot.password.title")</legend>
+        @helper.form(action = routes.ForgotPasswordController.submit(), 'autocomplete -> "off") {
```
Comment (misaugstad) [inline app/views/forgotPassword.scala.html:8]: Do we need to turn off autocomplete for the email address?

Comment (ThatOneGoat) [reply app/views/forgotPassword.scala.html:8]: Probably not? I can't think of any safety issues

File: app/models/services/AuthTokenServiceImpl.scala (hunk 1 above)
Comment (ThatOneGoat) [reply app/models/services/AuthTokenServiceImpl.scala:22]: Nope, I think I left those over

File: app/utils/di/SilhouetteModule.scala (hunk 2 above)
Comment (ThatOneGoat) [reply app/utils/di/SilhouetteModule.scala:48]: No, I switched to using Java.sql.Timestamp and Instant

File: app/models/daos/slick/UserDAOSlick.scala [hunk 4]
```diff
@@ -115,12 +115,14 @@ class UserDAOSlick extends UserDAO {
           case Some(info) => Logger.debug("Nothing to insert since info already exists: " + info)
         }
         dbLoginInfo = slickLoginInfos.filter(info => info.providerID === dbLoginInfo.providerID && info.providerKey === dbLoginInfo.providerKey).first
+        val dbUserLoginInfo = DBUserLoginInfo(dbUser.userId, dbLoginInfo.id.get)
         // Now make sure they are connected
-        slickUserLoginInfos.filter(info => info.userID === dbUser.userId && info.loginInfoId === dbLoginInfo.id).firstOption match {
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/slick/UserDAOSlick.scala:119]: Before, if would check if both the login info id and user id were the same. The problem is when a user created an account, the user_id would match to their old anon account, but the login_info_id would not be the same. In the context of this check, they would be considered different user_login_infos even with the same user_id. In hindsight, I think they person who wrote this meant || not &&.

File: app/models/daos/slick/PasswordDAOSlick.scala [hunk 5]
```diff
@@ -57,4 +64,5 @@ class PasswordInfoDAOSlick extends DelegableAuthInfoDAO[PasswordInfo] {
       }
     }
   }
+
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/slick/PasswordDAOSlick.scala:67]: Ok

File: app/models/daos/slick/AuthTokenDAOSlick.scala [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Removes the token for the given ID.
+   *
+   * @param id The ID for which the token should be removed.
+   * @return A future to wait for the process to be completed.
+   */
+  def remove(id: UUID) = {
+    DB withSession { implicit session =>
+      Future.successful {
+        slickAuthTokens.filter(x => x.id === id.toString).delete
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/slick/AuthTokenDAOSlick.scala:81]: Ok

File: app/models/daos/slick/AuthTokenDAOSlick.scala [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+   * Saves a token.
+   *
+   * @param token The token to save.
+   * @return The saved token.
+   */
+  def save(token: AuthToken) = {
+    DB withSession { implicit session =>
+      Future.successful {
+        val dbAuthToken = DBAuthToken(token.id.toString, token.userID.toString, token.expiry)
+        slickAuthTokens.filter(x => x.userID === dbAuthToken.userID).firstOption match {
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/slick/AuthTokenDAOSlick.scala:62]: Ok

File: app/models/daos/slick/AuthTokenDAOSlick.scala [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+class AuthTokenDAOSlick extends AuthTokenDAO {
+  import play.api.Play.current
+
+  /**
+   * Finds a token by its ID.
+   *
+   * @param id The unique token ID.
+   * @return The found token or None if no token for the given ID could be found.
+   */
+  def find(id: UUID) = {
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/slick/AuthTokenDAOSlick.scala:23]: Ok

File: conf/evolutions/default/79.sql [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# --- !Ups
+CREATE TABLE auth_tokens (
+  user_id character varying(254) NOT NULL,
+  id character varying(254) NOT NULL,
+  timestamp TIMESTAMPTZ
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply conf/evolutions/default/79.sql:5]: got it

File: app/models/daos/slick/AuthTokenDAOSlick.scala [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    DB withSession { implicit session =>
+      Future.successful {
+        slickAuthTokens.filter(
+          x => x.id === id.toString
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/slick/AuthTokenDAOSlick.scala:27]: ok

File: app/models/daos/AuthTokenDAOImpl.scala [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ */
+class AuthTokenDAOImpl extends AuthTokenDAO {
+
+  /**
+   * Finds a token by its ID.
+   *
+   * @param id The unique token ID.
+   * @return The found token or None if no token for the given ID could be found.
+   */
+  def find(id: UUID) = Future.successful(tokens.get(id))
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/AuthTokenDAOImpl.scala:25]: ok

File: app/controllers/ForgotPasswordController.scala [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        val loginInfo = LoginInfo(CredentialsProvider.ID, email)
+        val result = Redirect(routes.UserController.signIn()).flashing("info" -> Messages("email.reset.password.sent"))
+        userService.retrieve(loginInfo).flatMap {
+          case Some(user) =>
+            authTokenService.create(user.userId).map { authToken =>
+              val url = routes.UserController.resetPassword(authToken.id).absoluteURL()
+
+              val resetEmail = Email(
+                Messages("email.reset.title"),
+                "Project Sidewalk <mlprojectsidewalk@gmail.com>",
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/controllers/ForgotPasswordController.scala:51]: The first line is the email subject (from the Messages) and the second line is the "from" sender that the user will see (for some reason the from sender can be different than the actual user account configured for smtp? Not sure why...)

File: app/controllers/ForgotPasswordController.scala [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+              MailerPlugin.send(resetEmail)
+              result
+            }
+
+          case None => Future.successful(result)
+        }
+      }
+    )
+  }
+}
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/controllers/ForgotPasswordController.scala:65]: ok

**Comment by ThatOneGoat** [comment]
Ok, just a few questions. 

For 3), I'm not quite sure what you mean. I have that done on my local machine, but I'm assuming I probably don't want to push what I have since that's the actual address/password. Do you mean just push what I have in docker-compose.yml but just replace the address and password with something like "dummyaddress" and "dummypassword"?

For 4), just to clarify, what safety benefits does hashing the token have? Currently, all the tokens have are user_id, token_id, and expiration timestamp. I'm thinking that, if the database got compromised, hashing would prevent someone from changing the expiration timestamp and then freely changing a users password through the compromised token, but just want to make sure that's what you mean.

For 9), I'm not quite sure what you mean. Do you mean like the button that's like "Didn't receive an email? Press to send again?"

**Comment by misaugstad** [comment]
>Do you mean just push what I have in docker-compose.yml but just replace the address and password with something like "dummyaddress" and "dummypassword"?

Yep! 

>what safety benefits does hashing the token have? Currently, all the tokens have are user_id, token_id, and expiration timestamp. I'm thinking that, if the database got compromised, hashing would prevent someone from changing the expiration timestamp and then freely changing a users password through the compromised token, but just want to make sure that's what you mean.

Basically. If they can read the tokens (even without write access) then they can go to reset password URL and reset their password.

>For 9), I'm not quite sure what you mean. Do you mean like the button that's like "Didn't receive an email? Press to send again?"

It doesn't even need to be a separate button. Even if the user enters the same email and clicks the send email button again, we should send them a second email. For the same reason that other websites have a "send again" button.

**Comment by ThatOneGoat** [comment]
Currently, when an email is sent to the user, it redirects them to the signIn page with an "email sent" popup. Should I change it so it redirects to the reset password page again. From there, the ability for the user to just send another email to themselves should already be implemented.

**Comment by misaugstad** [comment]
No I think if they just click "forgot password" again at some point and end back up at the forgot password page and click send email it should send them an email. I thought that that wasn't happening when I tried to do that while testing.

**Comment by ThatOneGoat** [comment]
Wait, I'm still not sure what you mean. Right now what's happening is just whenever the user goes to the forgot password page, they need to input an email and click send email to have an email sent. I think right now it doesn't allow duplicates since each time an email is sent, a new token is made, but a user can only have one valid auth Token at a time (to prevent spamming). Do we want to enable letting users have multiple emails sent?

**Comment by misaugstad** [comment]
I think if they try and go through it again, don't make a new token, just resend the old token. To account for them never receiving the email or something like that.

**Comment by ThatOneGoat** [comment]
Should I create a separate table for password request logging or can I just use the existing WebpageActivity Table?

**Comment by ThatOneGoat** [comment]
> I think if they try and go through it again, don't make a new token, just resend the old token. To account for them never receiving the email or something like that.

I'm wondering if maybe we shouldn't resend the old token. Right now all we are storing is a hashed version of the token for security in the database, but if we were to send to old one, we would have to somehow store an unhashed version of that token too. 

I think it'd be safer to just replace the old token with a new token each time the user requests to send an email. Thoughts?

**Comment by misaugstad** [comment]
Yeah you’re right that would be better. Then I guess we either need to delete the old token when we create a new one, or when they reset we need to make sure to delete all the tokens we generated for them.

**Comment by ThatOneGoat** [comment]
Yeah, I'm just replacing the old one with the new token for now.

Another thing I'm wondering is how secure should the hash be? Right now I'm just using SHA-256 for speed but it's not necessarily the safest. Our passwords are currently being hashed with bcrypt but I know bcrypt isn't the fastest hashing algorithm, so which one would be better?

**Comment by misaugstad** [comment]
Well it depends on how long hashing with each of them takes. If it only takes a few ms anyway, then we might as well go with the faster one. If the hashing itself takes a full second for the slow one, then I'd say performance is more important at that point.

**Comment by ThatOneGoat** [comment]
It seems like for our case since the hashed part of the token is a randomly generated UUID and it's being stored temporarily, we should be ok just using SHA-256

**Comment by misaugstad** [comment]
Sounds good. Also want to clarify: maybe it was clear from context, but I meant to say "If it only takes a few ms anyway, then we might as well go with the more secure one.", but I said "go with the faster one". I still trust your decision to use the faster one if you think it makes sense in this context.

**Comment by ThatOneGoat** [comment]
Oh ye, I got that. I'm still thinking about whether it might be safer to use bcrypt tho since it seems it doesn't seem like there's a huge time difference (but that might just be my computer being weird). I'll give it some more thought.

File: app/models/services/AuthTokenService.scala [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+  /**
+   * Cleans expired tokens.
+   *
+   * @return The list of deleted tokens.
+   */
+  def clean: Future[Unit]
+
+  /**
+   * Remove token associated with given token id
```
Comment (misaugstad) [inline app/models/services/AuthTokenService.scala:41]: Missing a period at the end of this comment :upside_down_face:

File: app/models/services/AuthTokenServiceImpl.scala [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    }
+  }
+
+  /**
+   * Validates a token ID.
+   *
+   * @param id The token ID to validate.
+   * @return The token if it's valid, None otherwise.
+   */
+  def validate(id: UUID) = authTokenDAO.find(id).flatMap {
```
Comment (misaugstad) [inline app/models/services/AuthTokenServiceImpl.scala:53]: If the function is longer than one line, make sure to have it in this format:
```
  def validate(id: UUID) = {
    ...
  }
```
[Reference](https://docs.scala-lang.org/style/declarations.html)

File: app/models/services/AuthTokenServiceImpl.scala [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+  /**
+   * Cleans expired tokens.
+   *
+   * @return The list of deleted tokens.
+   */
+  def clean = authTokenDAO.removeExpired(new Timestamp(Instant.now.toEpochMilli))
+
+  /**
+   * Remove token associated with given token id
```
Comment (misaugstad) [inline app/models/services/AuthTokenServiceImpl.scala:69]: Missing period at end of this comment.

File: app/utils/actor/AuthTokenCleanerActor.scala [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+class AuthTokenCleanerActor extends Actor {
+
+  private var cancellable: Option[Cancellable] = None
+  val TIMEZONE = TimeZone.getTimeZone("UTC")
+  val authTokenService : AuthTokenService = new AuthTokenServiceImpl(new AuthTokenDAOSlick())
+
+  override def preStart(): Unit = {
+    super.preStart()
+
+    // If we want to update the user_stat table at 2:00am every day, we need to figure out how much time there
```
Comment (misaugstad) [inline app/utils/actor/AuthTokenCleanerActor.scala:28]: Can you change this to not say "user_stat table"

File: app/models/daos/slick/AuthTokenDAOSlick.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (misaugstad) [reply app/models/daos/slick/AuthTokenDAOSlick.scala:31]: Ahh I see, it extends `AuthTokenDAO` so you shouldn't be adding return types :+1:

**Comment by ThatOneGoat** [comment]
> 2\. If bcrypt is not taking much longer at all and it is just as simple to use, then I'd suggest switching to that. I'm not surprised that it doesn't take too long, we are only hashing a short string.

I'm pretty confident I actually messed up my implementation of bcrypt. It should take way longer than my tests have since it's not a short string that I'm hashing but a UUID. With the current password hasher (which also uses bcrypt), a 7 digit long password takes longer to hash than my implementation hashing a UUID.

I'm going to try to see if there's an existing package, but I expect it to be much slower since bcrypt is supposed to be an intentionally slow hashing function

File: app/models/services/AuthTokenService.scala [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+trait AuthTokenService {
+
+  /**
+   * Creates a new auth token and saves it in the backing store.
+   *
+   * @param userID The user ID for which the token should be created.
+   * @param expiry The duration a token expires.
+   * @return The saved auth token id.
+   */
+  def create(userID: UUID, expiry: FiniteDuration = 5 minutes): Future[UUID]
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/services/AuthTokenService.scala:23]: Yup, I'll change it to 60 min

File: app/models/services/AuthTokenServiceImpl.scala [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+  def sha256Hasher: MessageDigest = MessageDigest.getInstance("SHA-256")
+
+  /**
+   * Creates a new auth token and saves it in the backing store.
+   *
+   * @param userID The user ID for which the token should be created.
+   * @param expiry The duration a token expires.
+   * @return The saved auth token.
+   */
+  def create(userID: UUID, expiry: FiniteDuration = 5 minutes) = {
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/services/AuthTokenServiceImpl.scala:38]: Same as above

File: app/forms/ResetPasswordForm.scala [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+    )(PasswordData.apply)(PasswordData.unapply)
+  )
+
+  /**
+   * The password data.
+   *
+   * @param password The new password of the user.
+   * @param passwordConfirm The confirmed new password of the user
+   */
+  case class PasswordData(
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/forms/ResetPasswordForm.scala:27]: Got it

File: app/models/AuthToken.scala [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+import org.joda.time.DateTime
+
+/**
+ * A token to authenticate a user against an endpoint for a short time period.
+ *
+ * @param id The unique token ID.
+ * @param userID The unique ID of the user the token is associated with.
+ * @param expiry The date-time the token expires.
+ */
+case class AuthToken(
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/AuthToken.scala:15]: Got it

File: app/models/daos/slick/PasswordDAOSlick.scala [hunk 22]
```diff
@@ -17,11 +17,12 @@ class PasswordInfoDAOSlick extends DelegableAuthInfoDAO[PasswordInfo] {
   import play.api.Play.current
 
   /**
-   * Saves the password info.
+   * Updates the password info if user password info already exists, otherwise
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/slick/PasswordDAOSlick.scala:20]: Got it

File: conf/evolutions/default/79.sql [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+  expiration_timestamp TIMESTAMPTZ NOT NULL
+);
+
+DELETE FROM user_login_info T1
+USING user_login_info T2
+WHERE T1.ctid < T2.ctid
+AND T1.user_id = T2.user_id;
+
+# --- !Downs
+DROP TABLE auth_tokens;
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply conf/evolutions/default/79.sql:14]: Got it

File: docker-compose.yml [hunk 24]
```diff
@@ -15,7 +15,9 @@ services:
     ports:
       - "9000:9000"
     environment:
-      - DATABASE_URL=jdbc:postgresql://db:5432/sidewalk
+      - DATABASE_URL=jdbc:postgresql://db:5432/sidewalk-newberg
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply docker-compose.yml:18]: Yup

File: app/models/daos/AuthTokenDAO.scala [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * Finds a token by its ID.
+   *
+   * @param id The unique token ID.
+   * @return The found token or None if no token for the given ID could be found.
+   */
+  def find(id: UUID): Future[Option[AuthToken]]
+
+  /**
+   * Finds expired tokens.
```
[the remark this answers is not in the document]
Comment (ThatOneGoat) [reply app/models/daos/AuthTokenDAO.scala:26]: Got it

File: app/models/services/AuthTokenService.scala (hunk 14 above)
Comment (ThatOneGoat) [reply app/models/services/AuthTokenService.scala:41]: Yup

File: app/models/services/AuthTokenServiceImpl.scala (hunk 15 above)
Comment (ThatOneGoat) [reply app/models/services/AuthTokenServiceImpl.scala:53]: Got it

File: app/models/services/AuthTokenServiceImpl.scala (hunk 16 above)
Comment (ThatOneGoat) [reply app/models/services/AuthTokenServiceImpl.scala:69]: Yup

File: app/utils/actor/AuthTokenCleanerActor.scala (hunk 17 above)
Comment (ThatOneGoat) [reply app/utils/actor/AuthTokenCleanerActor.scala:28]: Got it

**Comment by ThatOneGoat** [comment]
> 1. Can we also log when a password reset _request_ occurs as well? We would probably want to include the email address they are trying to reset with when logging that as well...

If I make a different table for this, should I keep the actual "password reset" action in the WebpageActivityTable, make a separate table for password reset, or include the actual password rest action logging with the password reset request action table?

**Comment by misaugstad** [comment]
I think we should keep both actions in the webpage_activity table. You can just include the email address in the text of the action, similar to how the clicks on the choropleths log a bunch of info in the action.

**Comment by misaugstad** [comment]
Okay I tested and just made two tiny edits myself so that it's now ready to merge!
1. I changed "PasswordResetRequest to email" to "PasswordResetRequest=email"
1. The body of the email was not being translated to Spanish, so I fixed that. Just needed to add `(implicit lang: Lang)` to the end of the `@(user: User, url: String)` line in resetPasswordEmail.scala.

You did a really great job with this PR @ThatOneGoat so excited to finally merge it!!

**Pull request merged**